syntax/parser_arithm.go

```diff
@@ -0,0 +1,318 @@
+			return value
+		}
+		if !isArithName(value) {
+			p.posErr(p.pos, "%s must follow a name", p.tok.String())
+		}
+		pos := p.pos
+		tok := p.tok
+		p.nextArithOp(compact)
+		y := p.arithmExprAssign(compact)
+		if y == nil {
+			p.followErrExp(pos, tok.String())
+		}
+		return &BinaryArithm{
+			OpPos: pos,
+			Op:    BinAritOperator(tok),
+			X:     value,
+			Y:     y,
+		}
+	}
+	return value
+}
+
+func (p *Parser) arithmExprCond(compact bool) ArithmExpr {
+	value := p.arithmExprLor(compact)
+	if BinAritOperator(p.tok) == TernQuest {
+		if compact && p.spaced {
+			return value
+		}
+		if value == nil {
+			p.curErr("%s must follow an expression", p.tok.String())
+		}
+		questPos := p.pos
+		p.nextArithOp(compact)
+		if BinAritOperator(p.tok) == TernColon {
+			p.followErrExp(questPos, TernQuest.String())
+		}
+		trueExpr := p.arithmExpr(compact)
+		if trueExpr == nil {
+			p.followErrExp(questPos, TernQuest.String())
+		}
+		if BinAritOperator(p.tok) != TernColon {
+			p.posErr(p.pos, "ternary operator missing : after ?")
+		}
+		colonPos := p.pos
+		p.nextArithOp(compact)
+		falseExpr := p.arithmExprCond(compact)
+		if falseExpr == nil {
+			p.followErrExp(colonPos, TernColon.String())
+		}
+		return &BinaryArithm{
+			OpPos: questPos,
+			Op:    BinAritOperator(TernQuest),
+			X:     value,
+			Y: &BinaryArithm{
+				OpPos: colonPos,
+				Op:    BinAritOperator(TernColon),
+				X:     trueExpr,
+				Y:     falseExpr,
+			},
+		}
+	}
+
+	return value
+}
+
+func (p *Parser) arithmExprLor(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprLand, OrArit)
+}
+
+func (p *Parser) arithmExprLand(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprBor, AndArit)
+}
+
+func (p *Parser) arithmExprBor(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprBxor, Or)
+}
+
+func (p *Parser) arithmExprBxor(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprBand, Xor)
+}
+
+func (p *Parser) arithmExprBand(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExpr5, And)
+}
+
+func (p *Parser) arithmExpr5(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExpr4, Eql, Neq)
+}
+
+func (p *Parser) arithmExpr4(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprShift, Lss, Gtr, Leq, Geq)
+}
+
+func (p *Parser) arithmExprShift(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExpr3, Shl, Shr)
+}
+
+func (p *Parser) arithmExpr3(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExpr2, Add, Sub)
+}
+
+func (p *Parser) arithmExpr2(compact bool) ArithmExpr {
+	return p.arithmExprBinary(compact, p.arithmExprPower, Mul, Quo, Rem)
+}
+
+func (p *Parser) arithmExprPower(compact bool) ArithmExpr {
+	// Power is different from the other binary operators because it's right-associative
+	value := p.arithmExpr1(compact)
+	if BinAritOperator(p.tok) == Pow {
```

Comment:
if you invert this comparison, you can then unindent most of the code. the same applies to other funcs.